File: dawn_ocaml/lib/func_rb.ml
type color = Red | Black

type 'a rbtree =
  | Empty
  | Node of color * 'a rbtree * 'a * 'a rbtree

let balance = function
  | Black, Node (Red, Node (Red, a, x, b), y, c), z, d
  | Black, Node (Red, a, x, Node (Red, b, y, c)), z, d
  | Black, a, x, Node (Red, Node (Red, b, y, c), z, d)
  | Black, a, x, Node (Red, b, y, Node (Red, c, z, d)) ->
    Node (Red, Node (Black, a, x, b), y, Node (Black, c, z, d))
  | color, a, x, b -> Node (color, a, x, b)

let insert x s =
  let rec ins = function
    | Empty -> Node (Red, Empty, x, Empty)
    | Node (color, a, y, b) ->
       if x < y
       then balance (color, ins a, y, b)
       else if x > y
       then balance (color, a, y, ins b)
       else Node (color, a, y, b)
  in
  match ins s with
  | Node (_, a, x, b) -> Node (Black, a, x, b)
  | Empty -> Empty

File: dawn_haskell/src/Func_rb.hs
data Color = Red | Black deriving (Eq, Show)
data RBTree a = Empty | Node Color (RBTree a) a (RBTree a) deriving (Eq, Show)

balance :: Color -> RBTree a -> a -> RBTree a -> RBTree a
balance Black (Node Red (Node Red a x b) y c) z d =
  Node Red (Node Black a x b) y (Node Black c z d)
balance Black (Node Red a x (Node Red b y c)) z d =
  Node Red (Node Black a x b) y (Node Black c z d)
balance Black a x (Node Red (Node Red b y c) z d) =
  Node Red (Node Black a x b) y (Node Black c z d)
balance Black a x (Node Red b y (Node Red c z d)) =
  Node Red (Node Black a x b) y (Node Black c z d)
balance color a x b = Node color a x b

insert :: Ord a => a -> RBTree a -> RBTree a
insert x Empty = Node Red Empty x Empty
insert x (Node color a y b)
  | x < y = balance color (insert x a) y b
  | x > y = balance color a y (insert x b)
  | otherwise = Node color a y b